11-applicative2/SExpr.hs

```diff
@@ -0,0 +1,180 @@
+Just ("foobar"," baz")
+*AParser> runParser ident "foo33fA"
+Just ("foo33fA","")
+*AParser> runParser ident "2bad"
+Nothing
+*AParser> runParser ident ""
+Nothing
+-}
+------------------------------------------------------------
+--  2. Utilities
+------------------------------------------------------------
+
+spaces :: Parser String
+spaces = zeroOrMore $ satisfy isSpace
+
+ident :: Parser String
+ident = (:) <$> satisfy isAlpha <*> zeroOrMore (satisfy isAlphaNum)
+
+
+{-
+Exercise 3
+S-expressions are a simple syntactic format for tree-structured data,
+originally developed as a syntax for Lisp programs. We’ll close out
+our demonstration of parser combinators by writing a simple Sexpression parser.
+An identifier is represented as just a String; the format for valid
+identifiers is represented by the ident parser you wrote in the previous exercise.
+type Ident = String
+An “atom” is either an integer value (which can be parsed with
+posInt) or an identifier.
+data Atom = N Integer | I Ident
+deriving Show
+Finally, an S-expression is either an atom, or a list of S-expressions.
+
+data SExpr = A Atom
+| Comb [SExpr]
+deriving Show
+Textually, S-expressions can optionally begin and end with any
+number of spaces; after throwing away leading and trailing spaces they
+consist of either an atom, or an open parenthesis followed by one or
+more S-expressions followed by a close parenthesis.
+atom ::= int
+    | ident
+
+S ::= atom
+    | (S∗)
+For example, the following are all valid S-expressions:
+5
+foo3
+(bar (foo) 3 5 874)
+(((lambda x (lambda y (plus x y))) 3) 5)
+( lots of ( spaces in ) this ( one ) )
+We have provided Haskell data types representing S-expressions in
+SExpr.hs. Write a parser for S-expressions, that is, something of type
+parseSExpr :: Parser SExpr
+
+Hints: To parse something but ignore its output, you can use the
+(*>) and (<*) operators, which have the types
+(*>) :: Applicative f => f a -> f b -> f b
+(<*) :: Applicative f => f a -> f b -> f a
+p1 *> p2 runs p1 and p2 in sequence, but ignores the result of
+p1 and just returns the result of p2. p1 <* p2 also runs p1 and p2 in
+sequence, but returns the result of p1 (ignoring p2’s result) instead.
+For example:
+*AParser> runParser (spaces *> posInt) " 345"
+Just (345,"")
+-}
+------------------------------------------------------------
+--  3. Parsing S-expressions
+------------------------------------------------------------
+
+-- An "identifier" is represented as just a String; however, only
+-- those Strings consisting of a letter followed by any number of
+-- letters and digits are valid identifiers.
+type Ident = String
+
+-- An "atom" is either an integer value or an identifier.
+data Atom = N Integer | I Ident
+  deriving Show
+
+-- An S-expression is either an atom, or a list of S-expressions.
+data SExpr = A Atom
+           | Comb [SExpr]
+  deriving Show
+
+parseAtom :: Parser Atom
+parseAtom = N <$> posInt <|> I <$> ident
```

Comment:
The above does not make evaluation order obvious at a glance, we have to look up `infix` operator precedence (`<$>`, `<|>`) to know which evaluates first:
```haskell
parseAtom = N <$> posInt <|> I <$> ident
```

Something like this will be better, since we know prefix will get evaluated before infix parts:

```haskell
parseAtom = fmap N posInt <|> fmap I ident
```

Reply:
@kwannoel Thanks, 👍